File: app/lib/core/planner_classes.dart
import '../models/plan.dart';
import '../models/session.dart';
import 'semester_schedule.dart';

const Map<String, String> _kindLabel = {
  'lecture': 'L',
  'tutorial': 'T',
  'lab': 'Lab',
};

/// A planner class instance resolved onto a calendar date.
class PlannerClass {
  final String id;
  final String courseCode;
  final String kind; // lecture|tutorial|lab
  final String kindLabel;
  final String start;
  final String end;
  final String timeLabel; // HH:MM

  const PlannerClass({
    required this.id,
    required this.courseCode,
    required this.kind,
    required this.kindLabel,
    required this.start,
    required this.end,
    required this.timeLabel,
  });
}

String _formatSessionTime(String start) {
  if (start.length != 4) return start;
  return '${start.substring(0, 2)}:${start.substring(2)}';
}

/// Classes that run on [date] given the plan, honouring the academic calendar
/// (swaps/holidays). Mirrors `getClassesForDate` in `plannerClasses.js`.
List<PlannerClass> getClassesForDate(
  DateTime date,
  List<SelectedCourse> courses,
  Map<String, CourseTimetable> timetableData,
) {
  final academic = getAcademicDay(date);
  if (!academic.hasClasses || academic.effectiveDay == null) return const [];

  final sessions = <PlannerClass>[];
  for (final course in courses) {
    final data = timetableData[course.courseCode];
    if (data == null) continue;

    void addFrom(List<Session>? slots, String kind) {
      if (slots == null) return;
      for (final slot in slots) {
        if (slot.day != academic.effectiveDay) continue;
        sessions.add(PlannerClass(
          id: '${course.courseCode}-$kind-${slot.start}-${slot.end}',
          courseCode: course.courseCode,
          kind: kind,
          kindLabel: _kindLabel[kind] ?? kind,
          start: slot.start,
          end: slot.end,
          timeLabel: _formatSessionTime(slot.start),
        ));
      }
    }

    addFrom(data.lecture, 'lecture');
    addFrom(data.tutorial, 'tutorial');
    addFrom(data.lab, 'lab');
  }

  sessions.sort((a, b) {
    final c = a.start.compareTo(b.start);
    return c != 0 ? c : a.courseCode.compareTo(b.courseCode);
  });
  return sessions;
}

/// Map of dateKey -> classes for a list of dates (only dates with classes).
Map<String, List<PlannerClass>> buildClassesByDate(
  List<DateTime> dates,
  List<SelectedCourse> courses,
  Map<String, CourseTimetable> timetableData,
) {
  final map = <String, List<PlannerClass>>{};
  for (final date in dates) {
    final classes = getClassesForDate(date, courses, timetableData);
    if (classes.isNotEmpty) {
      map[formatDateKey(date)] = classes;
    }
  }
  return map;
}

File: src/utils/plannerClasses.js
// Reads the weekly planner from localStorage and resolves which sessions
// run on a given calendar date (respecting holidays, swaps and breaks via
// getAcademicDay).

import { formatDateKey, getAcademicDay } from './semesterSchedule';

const SESSION_KINDS = ['lecture', 'tutorial', 'lab'];

const KIND_LABELS = {
    lecture: 'L',
    tutorial: 'T',
    lab: 'Lab',
};

export function loadPlannerState() {
    try {
        const coursesRaw = localStorage.getItem('selectedCourses');
        const timetableRaw = localStorage.getItem('timetableData');
        const courses = coursesRaw ? JSON.parse(coursesRaw) : [];
        const timetableData = timetableRaw ? JSON.parse(timetableRaw) : {};
        if (!Array.isArray(courses)) {
            return { courses: [], timetableData: {} };
        }
        return { courses, timetableData };
    } catch (e) {
        return { courses: [], timetableData: {} };
    }
}

export function formatSessionTime(hhmm) {
    if (!hhmm || hhmm.length !== 4) return '';
    return `${hhmm.slice(0, 2)}:${hhmm.slice(2)}`;
}

/** Sessions on a calendar date from the planner, sorted by start time. */
export function getClassesForDate(date, courses, timetableData) {
    const academic = getAcademicDay(date);
    if (!academic.hasClasses || !academic.effectiveDay) return [];

    const sessions = [];
    for (const course of courses) {
        if (!course || !course.courseCode) continue;
        const data = timetableData[course.courseCode];
        if (!data) continue;

        for (const kind of SESSION_KINDS) {
            const slots = data[kind];
            if (!Array.isArray(slots)) continue;
            for (const slot of slots) {
                if (!slot || slot.day !== academic.effectiveDay) continue;
                // Match Plan timetable: per-slot location, else course lectureHall for lectures.
                const location =
                    (slot.location && String(slot.location).trim())
                    || (kind === 'lecture' && course.lectureHall
                        ? String(course.lectureHall).trim()
                        : '')
                    || '';
                sessions.push({
                    id: `${course.courseCode}-${kind}-${slot.start}-${slot.end}`,
                    courseCode: course.courseCode,
                    kind,
                    kindLabel: KIND_LABELS[kind],
                    start: slot.start,
                    end: slot.end,
                    timeLabel: formatSessionTime(slot.start),
                    location,
                });
            }
        }
    }

    sessions.sort((a, b) => {
        if (a.start !== b.start) return a.start < b.start ? -1 : 1;
        return a.courseCode.localeCompare(b.courseCode);
    });
    return sessions;
}

/** Map YYYY-MM-DD → session[] for every date in `dates` (Date objects). */
export function buildClassesByDate(dates, courses, timetableData) {
    const map = new Map();
    for (const date of dates) {
        const key = formatDateKey(date);
        const sessions = getClassesForDate(date, courses, timetableData);
        if (sessions.length > 0) map.set(key, sessions);
    }
    return map;
}
